pkg/core/tokens/signing_key_accessor.go

```diff
@@ -26,43 +26,62 @@ type SigningKeyAccessor interface {
 type signingKeyAccessor struct {
 	resManager       manager.ResourceManager
 	signingKeyPrefix string
+	isPublic         bool
 }
 
 var _ SigningKeyAccessor = &signingKeyAccessor{}
 
-func NewSigningKeyAccessor(resManager manager.ResourceManager, signingKeyPrefix string) SigningKeyAccessor {
+func NewSigningKeyAccessor(resManager manager.ResourceManager, signingKeyPrefix string, isPublic bool) SigningKeyAccessor {
 	return &signingKeyAccessor{
 		resManager:       resManager,
 		signingKeyPrefix: signingKeyPrefix,
+		isPublic:         isPublic,
 	}
 }
 
 func (s *signingKeyAccessor) GetPublicKey(ctx context.Context, serialNumber int) (*rsa.PublicKey, error) {
-	keyBytes, err := s.getKeyBytes(ctx, serialNumber)
+	keyResource, err := s.getKey(ctx, serialNumber)
 	if err != nil {
 		return nil, err
 	}
-	key, err := keyBytesToRsaKey(keyBytes)
+
+	keyBytes := keyResource.Spec.GetData().GetValue()
+
+	if s.isPublic {
+		return keyBytesToRsaPublicKey(keyBytes)
+	}
+
+	key, err := keyBytesToRsaPrivateKey(keyBytes)
 	if err != nil {
 		return nil, err
 	}
+
 	return &key.PublicKey, nil
 }
 
-func (s *signingKeyAccessor) getKeyBytes(ctx context.Context, serialNumber int) ([]byte, error) {
+func (s *signingKeyAccessor) getKey(ctx context.Context, serialNumber int) (*system.GlobalSecretResource, error) {
+	resourceKey := SigningKeyResourceKey(s.signingKeyPrefix, serialNumber, model.NoMesh)
 	resource := system.NewGlobalSecretResource()
-	if err := s.resManager.Get(ctx, resource, store.GetBy(SigningKeyResourceKey(s.signingKeyPrefix, serialNumber, model.NoMesh))); err != nil {
+
+	if err := s.resManager.Get(ctx, resource, store.GetBy(resourceKey)); err != nil {
 		if store.IsResourceNotFound(err) {
 			return nil, &SigningKeyNotFound{
 				SerialNumber: serialNumber,
 				Prefix:       s.signingKeyPrefix,
 			}
 		}
+
 		return nil, errors.Wrap(err, "could not retrieve signing key")
 	}
-	return resource.Spec.GetData().GetValue(), nil
+
+	return resource, nil
 }
 
 func (s *signingKeyAccessor) GetLegacyKey(ctx context.Context, serialNumber int) ([]byte, error) {
-	return s.getKeyBytes(ctx, serialNumber)
+	keyResource, err := s.getKey(ctx, serialNumber)
```

Comment:
I don't think it's the correct implementation to return anything other than an error from this method when it comes to the new Zone Token.

To quote the doc for this method
```
	// GetLegacyKey returns legacy key. In pre 1.4.x version of Kuma, we used symmetric HMAC256 method of signing DP keys.
	// In that case, we have to retrieve private key even for verification.
```

but we don't have a private key at all.
You can add `if s.isPublic {` here, but I'd leave it as is and create another implementation of `SigningKeyAccessor` called something like `PublicKeyInSecretAccessor` (there has to be a better name) that support this exact use case that we have only a public key in secret.

Reply:
addressed